### PR TITLE
VM: Jump dest analysis enhancements

### PR DESCRIPTION
**Pull request by acolytec3**
- Switch valid Jumps storage from multiple `number[]` to a single Uint8Array of predefined length
- Only do jump destination analysis if `JUMP`/`JUMPI`/`JUMPSUB` opcode encountered

Should finish off #1575

### CODE REVIEW

**Comment by ryanio** [comment]
just had one comment but otherwise this lgtm! thanks for taking it on so quickly

let's wait for one more approval from @jochem-brouwer or @holgerd77 since this is critical code (I think we should expect for vm internals)

**Comment by acolytec3** [comment]
> just had one comment but otherwise this lgtm! thanks for taking it on so quickly
> 
What's the comment?  :-)

**Review by ryanio — CHANGES_REQUESTED** [review]

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:104]: I'm not sure in what order JS evaluates an `A && B` statement, but if A is first evaluated and then B then I'd say swap A and B in this case, or probably even better cache a boolean which saves if we have ran JUMPDEST analysis.

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:263]: This creates a byte array while Geth uses a bitmap (so 1/8 length of the array is used there). I'm not sure if this is faster or practical in JS but worth mentinoing here.

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:269]: This still does string comparison which per #1575 should be removed. 
```
    for (let i = 0; i < code.length; i++) {
      const curOpCode = code[i]

      // no destinations into the middle of PUSH
      if (curOpCode >= 0x60 && curOpCode <= 0x7f){
        i += curOpCode - 0x5f
      }
```

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:274]: String comparison

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:279]: String comparison

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [reply packages/vm/src/evm/interpreter.ts:104]: In fact I think we should use a cache, there is an edge case where there is a JUMP but there are no JUMPDESTs, in that case `validJumps.length` remains 0.

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:24]: Can we add a comment here what this Uint8Array means? It should describe that opcode at pc `index` has value 0 (default), 1 (jumpdest), 2 (beginsub)

File: packages/vm/src/evm/interpreter.ts
Comment (acolytec3) [reply packages/vm/src/evm/interpreter.ts:263]: I don't think the bitmaps thing is relevant to us.  Geth is flipping bits for every push (or flipping a bunch at once) where as we don't do any writes at all for `PUSH` but just skip over however many `PUSH`s there are.

File: packages/vm/src/evm/interpreter.ts
Comment (acolytec3) [reply packages/vm/src/evm/interpreter.ts]: Makes sense.  I've added a new property to `runState` to track this.

File: packages/vm/src/evm/interpreter.ts
Comment (acolytec3) [reply packages/vm/src/evm/interpreter.ts]: Good call, fixed.

File: packages/vm/src/evm/interpreter.ts
Comment (acolytec3) [reply packages/vm/src/evm/interpreter.ts]: Does the comment I added align with your thoughts?

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [inline packages/vm/src/evm/interpreter.ts:104]: Super nit but maybe we want to name this flag `shouldDoJumpAnalysis` which we initialize to `true`, this saves a NOT operation for each opcode.

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [reply packages/vm/src/evm/interpreter.ts:264]: Agreed

File: packages/vm/src/evm/interpreter.ts
Comment (jochem-brouwer) [reply packages/vm/src/evm/interpreter.ts]: Yep 👍 😄

**Review by jochem-brouwer — APPROVED** [review]
LGTM!

**Review by ryanio — APPROVED** [review]

**Review by jochem-brouwer — APPROVED** [review]

**Comment by jochem-brouwer** [comment]
🎉

**Pull request merged**
